src/date/parse.js

```diff
@@ -236,24 +251,43 @@ return function( value, tokens, properties ) {
 		return null;
 	}
 
-	if ( era === 0 ) {
-		// 1 BC = year 0
-		date.setFullYear( date.getFullYear() * -1 + 1 );
+	if ( ret.month == null && ret.date == null && truncateAt.indexOf( YEAR ) !== -1) {
+		// year was defined but month was not
+		gdate = new calendar( dateStartOf ( date, "year", gdate ) );
+		ret.month = gdate.getMonth();
+		ret.date = gdate.getDate();
+	}
+	if ( ret.month == null ) {
+		ret.month = gdate.getMonth();
+	}
+	if ( ret.date == null && truncateAt.indexOf( MONTH ) !== -1) {
+		// month was defined but date was not
+		ret.date = 1; // gdate assumes 1 is the first date of the month
+	}
+	if ( ret.date == null ) {
+		ret.date = gdate.getDate();
 	}
 
-	if ( day !== undefined ) {
-		if ( outOfRange( day, 1, dateLastDayOfMonth( date ) ) ) {
-			return null;
-		}
-		date.setDate( day );
-	} else if ( daysOfYear !== undefined ) {
-		if ( outOfRange( daysOfYear, 1, dateIsLeapYear( date.getFullYear() ) ? 366 : 365 ) ) {
+  gdate = new calendar(ret.era, ret.year, ret.month, ret.date);
+
+	if ( daysOfYear !== undefined ) {
+    startOfYear = dateStartOf(gdate.toDate(), "year", gdate);
+    gdate = new calendar( startOfYear ).nextDate(daysOfYear - 1);
+		if ( gdate.getYear() !== startOfYear.getFullYear() ) {
 			return null;
 		}
-		date.setMonth(0);
-		date.setDate( daysOfYear );
+	}else if ( gdate.getDate() !== ret.date ) {
+		// if the date was invalid and gdate corrected it (like turning 31 Sep into 30 Sep),
+		// we want to reject it.
+		return null;
 	}
 
+  date = gdate.toDate();
+  date.setHours(ret.hour);
+  date.setMinutes(ret.minute);
+  date.setSeconds(ret.second);
+  date.setMilliseconds(ret.milliseconds);
```

Comment:
Can we reverse this logic? I mean keep handling time fields the way we were before and then setting the date fields of `date` by using `setFns` functions with `gdate` values.